### ISSUE TITLE
[1.8.9 L5][Emissive][CIT] Item emissive textures don't work for CIT items

### ISSUE THREAD

**Issue by TBlazeWarriorT** [sp614x/optifine#3358]
Labels: [1.8.9][Emissive][CIT][Bug][Todo (lol)]

OptiFine 1.8.9 HD U L5

Short Summary:
Emissive textures (_e) will not work inside /minecraft/mcpatcher/cit/.

Proof screenshot below. The mob's emissive works properly, but the items on the item frames (and everywhere else) looks just dark. I triple-checked to make sure it had the right format, seems just straight up broken for CIT items inside the /mcpatcher folder (/optifine/cit doesn't work for 1.8.9-1.12)
![image](https://user-images.githubusercontent.com/25572454/74619305-d51e9080-5113-11ea-917a-046947ce1fe1.png)
This is a test resource pack that also shows it. It contains the custom squid with working emissive and the sword with broken emissive.
[optifinetest.zip](https://github.com/sp614x/optifine/files/4211554/optifinetest.zip)

Reproducing:
Add a CIT item inside /mcpatcher/cit/. Add an emissive (_e) version of the png (custom_item_e.png). Add emissive.properties inside /optifine/ (since /mcpatcher still grabs emissive from the /optifine folder). The CIT texture will work, _e won't.

Emissive block textures don't seem to work either but this requires more testing.

**Comment by zico1115** [comment]
is there a way to get it to work?

**Comment by TBlazeWarriorT** [comment]
Seems to also affect 3D armor. Armor models will not work with emissive textures (for example, something_layer_1_e.png will do nothing if inside the cit folder). This bug is fixed in optifine 1.14, but 1.8 cannot have fancy emissive textures, of course..

**Comment by TsukiTheDwagon** [comment]
Also happens on Optifine 1.8.9 HD U M5.

Would really like a fix for this...

**Comment by 182exe** [comment]
god please fix (bump)

**Comment by anzz1** [comment]
**TLDR; Fix**

1. Have "OptiFine 1.8.9 HD U M6_pre2" installed.
2. Open `%APPDATA%\.minecraft\libraries\optifine\OptiFine\1.8.9_HD_U_M6_pre2\OptiFine-1.8.9_HD_U_M6_pre2.jar` with WinRAR or 7-Zip.
3. Put the attached pre-patched "bgg.class" file in the JAR archive, replacing the original "bgg.class".
4. Enjoy ?

[Emissive-Texture-Fix-OF-1.8.9_HD_U_M6_pre2.zip](https://github.com/sp614x/optifine/files/10375343/Emissive-Texture-Fix-OF-1.8.9_HD_U_M6_pre2.zip) 

**Technical info**
The problem is caused by a race condition in the logic for loading textures for BakedQuads. BakedQuads are used in all simple "pseudo-3D" models which are generated from 2D sprites like most blocks and items.

The condition happens when texture is loaded *before* the emissive texture configuration. thus returning that there is no emissive texture available. For optimization, this check is then stored in a boolean flag where the texture in question is "already checked" that it's not emissive, so even after the configuration is loaded it will still return "no texture available".

Here is the bytecode for the problem function, I've marked the check in question with ( * )
```
 * 00000000    2A          aload_0
 * 00000001    B4 00 DF    getfield            boolean bgg.quadEmissiveChecked
 * 00000004    99 00 08    ifeq                pos.0000000C
 * 00000007    2A          aload_0
 * 00000008    B4 00 E1    getfield            bgg bgg.quadEmissive
 * 0000000B    B0          areturn
   0000000C    2A          aload_0
   0000000D    B4 00 E1    getfield            bgg bgg.quadEmissive
   00000010    C7 00 27    ifnonnull           pos.00000037
   00000013    2A          aload_0
   00000014    B4 00 25    getfield            bmi bgg.sprite
   00000017    C6 00 20    ifnull              pos.00000037
   0000001A    2A          aload_0
   0000001B    B4 00 25    getfield            bmi bgg.sprite
   0000001E    B4 00 E4    getfield            bmi bmi.spriteEmissive
   00000021    C6 00 16    ifnull              pos.00000037
   00000024    2A          aload_0
   00000025    BB 00 E6    new                 bgn
   00000028    59          dup
   00000029    2A          aload_0
   0000002A    2A          aload_0
   0000002B    B4 00 25    getfield            bmi bgg.sprite
   0000002E    B4 00 E4    getfield            bmi bmi.spriteEmissive
   00000031    B7 00 E9    invokespecial       void bgn.<init>(bgg, bmi)
   00000034    B5 00 E1    putfield            bgg bgg.quadEmissive
 * 00000037    2A          aload_0
 * 00000038    04          iconst_1
 * 00000039    B5 00 DF    putfield            boolean bgg.quadEmissiveChecked
   0000003C    2A          aload_0
   0000003D    B4 00 E1    getfield            bgg bgg.quadEmissive
   00000040    B0          areturn
```

Patching the marked bytes with 00 (nop aka "do nothing" in JVM bytecode) will make the emissive textures work again. Better fix though would be to make sure that the emissive config is loaded before any textures are, but this'll do for a quick fix. This particular optimization does not save any meaningful amount of CPU time anyway so skipping it should have no performance impact.

The sequence starts in "bgg.class" at offset 0x1532, so here's how to patch it yourself with a hex editor:
0x1532: `2A B4 00 DF 99 00 08 2A B4 00 E1 B0` -> `00 00 00 00 00 00 00 00 00 00 00 00`
0x1569: `2A 04 B5 00 DF` -> `00 00 00 00 00`

**Comment by jacktech431** [comment]
@anzz1 
Thank you for this. I did have on question:

Is the proper location for the patched bgg.class file in the patch folder within the jar file?
Here is the location:
.minecraft\libraries\optifine\OptiFine\preview_OptiFine_1.8.9_HD_U_M6_pre2.jar\patch\

There exists bgg.class.md5 and bgg.class.xdelta in that location which I attempted to replace yet both came back after launching the game.
Otherwise I could not find a the unpatched bgg.class file.

**Comment by anzz1** [comment]
@jacktech431 You are trying to modify the installer JAR, not the installed final product, that's the problem.

Somewhat confusingly the "preview_OptiFine_1.8.9_HD_U_M6_pre2.jar" you download from the optifine site is NOT the same as the installed "OptiFine-1.8.9_HD_U_M6_pre2.jar" file.

The one which you downloaded, that is not what you put in the libraries folder, but rather you run it which pops up this installer
![image](https://user-images.githubusercontent.com/13628128/218167310-276e8f89-5ac8-4456-9bd2-0c156d0b9434.png)

and *after* installing, find the right JAR file and replace the class file. There should be one called **exactly** "bgg.class" , as "bgg.class.md5" and "bgg.class.xdelta" are part of the installer JAR which is the wrong one.

**Comment by anzz1** [comment]
(Technically the .class file is created from the .xdelta file while installing so it is certainly possible to modify the installer itself and apply the fix that way too, but they are different formats doing different things so you cannot simply replace a .class file with a .xdelta file or vice versa)
